### ISSUE TITLE
Fly scanning support?

### ISSUE THREAD

**Issue by jadball** [FABLE-3DXRD/ImageD11#91]
Hello,

I was wondering whether ImageD11, in principle, supports a variable omega step between scans?
Our idea to speed up data collection is to perform a fly scan with our tomography stage and constantly poll the detector with a 1 second exposure such that each second of exposure equates to an omega step of around 0.2 deg.
Is a variable omega step supported by the peaksearcher when reading from header files?

**Comment by jonwright** [comment]
I think so, yes.

**Comment by jadball** [comment]
Thanks, I'll maybe write a quick script to check.
Do you think attempting a fly scan is a good idea? It looks like we will run into problems if we wanted to do strain refinement with FitAllB as it asks for a constant delta-omega in the input file...

**Comment by jonwright** [comment]
The delta-omega for fitallb is to estimate the omega error. It uses the omega from peaksearch and does not know about your images. If the step size is always similar then it doesn't matter.

Whether it is a good idea depends very much on your control software. My experience up to now is that it is best to measure what actually happens in the experiment and adapt to the existing software as best you can. 

Apart from peak positions, I think the main difficulty is to figure out the normalisation properly if you want to get accurate intensities. Depends how much jitter you have on the motor and beam. We have used data from this kind of scan in the past, I don't remember now how we got the omega data in, I think we just fitted the average step size over the range. It was cheaper than editing all the headers to put the right Omega values.

I guess you would prefer a new option to read a table with filename, motorpos on each line?

**Comment by jadball** [comment]
As we use the same rotation stages for fly-scan tomography, we're able to get accurate motor positions per scan. I think the difficulty is how to feed that info into peaksearch.

In an ideal world, yes it would be nice to be able to provide peaksearch with a linked .csv for the scan with the motor positions. I suppose the slop can be calculated by taking an average for delta-omega and dividing by 2.

Currently our data is written in the following way:
- Nexus (.nxs) file with all the metadata for the scan, including motor positions for each scan etc. This can be exported to a CSV using Dawn software
- Linked folder with individual .cbf files which have empty headers

I suppose as a workaround I could use fabio to write correct header data into the cbfs?

**Comment by jonwright** [comment]
It would make sense (to me) to read the nxs file and cbf directly, but the shortest path might be to add an Omega in the cbf headers for now.
I guess we should re-write / re-factor peaksearcher.py anyway. Threading models have changed since it was written anyway.

**Comment by jadball** [comment]
Sure thing. It's fairly easy with h5py to extract our motor positions from the NXS file:

```
file_path = "path/to/nxs/file.nxs"
myfile = h5py.File(file_path,"r")
mygroup = myfile["entry1/instrument/ss1_theta"]
mydset = mygroup.get("ss1_theta")
print(mydset[...,:])
```

**Comment by jadball** [comment]
I've just run this through with my supervisor. He's happy for me to help update peaksearcher.py, if you want an extra pair of eyes on this?

My idea would be to add a new format .nxs to the list. If selected, you would also need to provide the internal path within the NXS file to the omega dataset, and to the image paths, perhaps as other command line options with argparse.

There would be a check in peaksearch_driver for .nxs selected - it would then generate the file series from the NXS metadata and a header object for each image with omega

**Comment by jonwright** [comment]
I'd rather not bring Nexus into ImageD11 for now, ESRF is going over to a Nexus format and for the moment it is all a bit confusing. There are fundamentalists who would want to read the image data via the nxs for example. In the long run, I think it will be better to go via sparse images anyway and keep all the pixels inside the peaks, but that is still a bit of a work in progress.

I will push a small example showing how you call peaksearch without going via the command line. Can you modify that example to read your nxs and image frames and add it in as a pull request for the next person to find somewhere (sandbox or test) please?

Thanks!

**Comment by jonwright** [comment]
Here it is (sorry, I am not very good at git)

https://github.com/FABLE-3DXRD/ImageD11/blob/master/test/peaksearchtiftest/scriptedpeaksearch.py

**Comment by jadball** [comment]
Created a new script to allow multithreading etc. Raised a PR: #93

**Comment by jonwright** [comment]
Assuming this is fixed by #93, so closing for now. Doubtless it will come back when ESRF go to their new dataformat.
